cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Lookup.java

```diff
@@ -16,15 +16,43 @@
 
 package co.cask.cdap.etl.api;
 
+import co.cask.cdap.api.data.schema.Schema;
+
+import java.util.Collection;
 import java.util.Map;
 import java.util.Set;
+import javax.annotation.Nullable;
 
 /**
  * Exposes read-only lookup operations on datasets.
  *
- * @param <T> the type of object that will be returned for a lookup
+ * @param <T> the type of object that will be returned for a lookup with String as the key
+ * @param <R> the raw type of object that will be returned for lookups with byte[] as the key
  */
-public interface Lookup<T> {
+public interface Lookup<T, R> {
+
+  /**
+   * @return Schema of the dataset, or null if no schema is stored in the dataset's properties
+   */
+  @Nullable
+  Schema getSchema();
+
+  /**
+   * Performs a single lookup.
+   *
+   * @param key the key to lookup
+   * @return the value associated with the key
+   */
+  R lookup(byte[] key);
+
+  /**
+   * Performs a batch lookup.
+   *
+   * @param keys the keys to lookup
+   * @return a map from key to value
+   */
+  Map<byte[], R> lookup(byte[]... keys);
```

Comment:
I couldn't add that, because the method signature would be the same for the added method (simply a different parameter in the Set parameterization doesn't change the method signature).
We could rename the new method, but then it wouldn't be consistent anymore, and I didn't see a strong use case for such a method.